Add support for s3 fields in discover

ps48 opened the pull request:
### Description

* Add support for s3 fields in discover

  

### Issues Resolved

  * Added async loaders to field selectors in side bar
  * On page reload, fetch the fields for the selected datasets
  * Added support for SQL/PPL fields to OSD field types
  * Added field type conversion from SQL/PPL to OSD field types
  * Added areFieldsLoading flag to the index pattern specification to showcase fields loading in the side bar
  * Added supportsTimeFilter flag to datasetConfig and dataset meta

## Screenshot

https://github.com/user-attachments/assets/4a91edf2-5af6-49ec-977d-958992dfcd62


## Changelog

- fix: Discover 2.0 support for S3 fields

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

ashwin-pc commented on src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts:
```diff
@@ -92,25 +93,32 @@ export class DatasetService {
-      } as IndexPatternSpec;
-      const temporaryIndexPattern = await this.indexPatterns?.create(spec, true);
-      if (temporaryIndexPattern) {
-        this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
+    const type = this.getType(dataset?.type);
+    try {
+      if (dataset && dataset.type !== DEFAULT_DATA.SET_TYPES.INDEX_PATTERN) {
+        const fetchedFields = await type?.fetchFields(dataset);
+        const spec = {
+          id: dataset.id,
+          title: dataset.title,
+          timeFieldName: dataset.timeFieldName,
+          fields: fetchedFields ?? ({} as IndexPatternFieldMap),
+          areFieldsLoading: true,
+          dataSourceRef: dataset.dataSource
+            ? {
+                id: dataset.dataSource.id!,
+                name: dataset.dataSource.title,
+                type: dataset.dataSource.type,
+              }
+            : undefined,
+        } as IndexPatternSpec;
+        const temporaryIndexPattern = await this.indexPatterns?.create(spec, true);
+
+        if (temporaryIndexPattern) {
+          this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
+        }
       }
+    } catch (error) {
```
Why not something like this?

```ts
public async cacheDataset(dataset: Dataset): Promise<void> {
    const type = this.getType(dataset?.type);
    try {
      const asyncType = type && type.isAsync;
      if (dataset && dataset.type !== DEFAULT_DATA.SET_TYPES.INDEX_PATTERN) {
        const fetchedFields = asyncType ? await type?.fetchFields(dataset) : undefined;
        const spec = {
          id: dataset.id,
          title: dataset.title,
          timeFieldName: dataset.timeFieldName,
          fields: fetchedFields ?? ({} as IndexPatternFieldMap),
          schemaLoading: asyncType ? true : false,
          dataSourceRef: dataset.dataSource
            ? {
                id: dataset.dataSource.id!,
                name: dataset.dataSource.title,
                type: dataset.dataSource.type,
              }
            : undefined,
        } as IndexPatternSpec;
        const temporaryIndexPattern = await this.indexPatterns?.create(spec, true);

        // Load schema asynchronously if it's an async index pattern
        if (asyncType && temporaryIndexPattern) {
          type.fetchFields(dataset).then((fields) => {
            temporaryIndexPattern.fields = fields;
            temporaryIndexPattern.schemaLoading = false;
            this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
          });
        }

        if (temporaryIndexPattern) {
          this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
        }
      }
    } catch (error) {
      throw new Error(`Failed to load dataset: ${dataset?.id}`);
    }
  }
  ```

This keeps the fetching of the fields colocated with the dataset service. It also avoids the situation where the loading of the field need to be coorectly specified everywhere we want to add support for async index patterns

ps48 replied:
Nice, this is cleaner than my implementation. Earlier, I was afraid of adding lot of flags to the dataset code base. But this makes sense will update this bit.

ps48 replied:
Updated here: [6d6a9c6](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/6d6a9c6b6100e4ca696875e237fe793c07fd84a8)

ashwin-pc commented on src/plugins/discover/public/application/view_components/utils/use_index_pattern.ts:
```diff
@@ -41,6 +45,20 @@ export const useIndexPattern = (services: DiscoverViewServices) => {
     data.indexPatterns,
   ]);
 
+  const fetchDatasetFields = useCallback(
```
I think we should handle this in the index pattern service itself. HEre's what im thinking:
1. You create an index pattern with or without schema information depending on dataset type. in the config you can specify async for the type to determine which of the two it is
2. Create a temporary index pattern when the dataset is selected in the configurator. but mark the type as async and defer loading of the schema. This is when you cache the indexpattern with an empty schema
3. while caching, start a promise right after caching it in the index pattern service that when it resolves adds he mappping to the index patttern and set the state to false. See my suggestion in the dataset serivce for reference

kavilla replied:
+1. i think it puts the heavy lifting on the discover plugin. which when we were doing this before the dataset selector there were inaccuracies in the results when we let the plugins ij\nterchange information and try to populate the fields. imo it's a little bit safer for data accuracy if the temp index pattern was created with the fields as a lot of the logic within this area of the code expected the fields to be present if they exist, and non mutating. 

since we update the current index pattern every time the dataset changes this is likely causing an issue where it set index the index pattern and if this is called any time the index pattern is set then its get fields again and sets the index pattern. which then becomes a race condition to which is the right query if the user in the meantime tries to update the language.

usually would be ok merging and fast follows. but the statefulness, multiple requests, and re-renders actually tend to be the source of a bugs that tend to blow up as of late.

ps48 replied:
+1 Updated the code to be async field fetch and not touch use_index_pattern hook.

ashwin-pc commented on src/plugins/query_enhancements/public/datasets/s3_type.ts:
```diff
@@ -247,3 +260,63 @@ const fetchTables = async (http: HttpSetup, path: DataStructure[]): Promise<Data
+
+    // Only include rows with valid types
+    if (dataType && dataType !== '') {
+      const sqlType = dataType as OPENSEARCH_SQL_FIELD_TYPES;
```
The SQL Types should be defined here

ps48 replied:
Updated here: [6d6a9c6](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/6d6a9c6b6100e4ca696875e237fe793c07fd84a8)

ashwin-pc reviewed (commented):
While most of the comments are abou code changes that dont affect users and can be fast followed, there are 2 user facing bugs that we need to address:
1. There are 3 requests fired when both fields are being loaded and the query is running. I can reproduce this the easiest when ive selected an s3 datasource and refreshed the page. See screenshot
<img width="661" alt="Screenshot 2024-10-18 at 4 51 33 AM" src="https://github.com/user-attachments/assets/092beb8e-4771-4788-9c3b-e7852b651a9d">

2. The available fields just dint stop spinning when the page refreshed even once the requests are successful. See secind screenshot
<img width="1751" alt="Screenshot 2024-10-18 at 5 23 29 AM" src="https://github.com/user-attachments/assets/d5eb27a3-f627-4095-8b55-f1bb9799a705">

virajsanghvi reviewed (commented):
didn't finish my review as I assume there will be changes based on ashwin's comments

ps48 commented:
Looks like Flint is breaking on main, get this when running any s3 based SQL query:

```
server    log   [21:53:31.717] [error][plugins][queryEnhancements] Facet fetch: enhancements.sqlQuery: Error: Data Source Error: Internal Server Error
server    log   [21:53:31.717] [error][plugins][queryEnhancements] sqlSearchStrategy: {
  "error": {
    "reason": "There was internal problem at backend",
    "details": "Glue storage engine is not supported.",
    "type": "UnsupportedOperationException"
  },
  "status": 500
}
```

ps48 commented:
Debugging the commits led me this https://github.com/opensearch-project/OpenSearch-Dashboards/commit/923cce847b09001dc5b3730c9bad54857d363532. this commit might be the root-cause why [we're seeing](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609#pullrequestreview-2377721289) this weird behavior for flint datasources

ps48 commented:
Current status: 
* Rebased my PR on main again - no conflicts as of now
* Reverted the dataselector [923cce8](https://github.com/opensearch-project/OpenSearch-Dashboards/commit/923cce847b09001dc5b3730c9bad54857d363532) changes on local to test my PR it works
* I am able to load and reload pages without any issue
* I the bug Ashwin raised on 3 queries being emitted out for fields -> Will take this and other non-blocking comments in fast follow once we get this merged in.

Video of local test with the revert dataselector changes: 

https://github.com/user-attachments/assets/79eafa5a-c5c4-4f2d-9978-c03037896102

kavilla commented on src/plugins/data/public/query/query_string/dataset_service/types.ts:
```diff
@@ -55,7 +58,7 @@ export interface DatasetTypeConfig {
    * @returns {Promise<DatasetField[]>} A promise that resolves to an array of DatasetFields.
    */
-  fetchFields: (dataset: Dataset) => Promise<DatasetField[]>;
+  fetchFields: (dataset: Dataset, http?: HttpSetup) => Promise<DatasetField[]>;
```
+1 we services

ps48 replied:
This came of Sean's comment, https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8337#discussion_r1786971503 will update it back to services.

ps48 replied:
Updated here: Updated here: [6d6a9c6](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/6d6a9c6b6100e4ca696875e237fe793c07fd84a8)

ps48 commented on src/plugins/data/common/osd_field_types/osd_field_types.ts:
```diff
@@ -84,3 +84,19 @@ export const setOsdFieldOverrides = (newOverrides: { [key: string]: any } | unde
+ *  @param {string} sqlType
+ *  @return {OSD_FIELD_TYPES}
+ */
+export function castSQLTypeToOSDFieldType(
```
Agree, will move this out of `osd_fields` altogether. This seems to be conflicting with earlier review from Rocky https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8337#discussion_r1776826366. In future, we can define clear tenets so that the contributors know the clear delineation between what goes in data plugin and what goes in query enhancement.

ps48 replied:
Updated here: [6d6a9c6](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/6d6a9c6b6100e4ca696875e237fe793c07fd84a8)

ps48 commented on src/plugins/discover/public/application/components/sidebar/discover_sidebar.tsx:
I use this config: 
```
  "[typescript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode",
    "editor.codeActionsOnSave": {
      "source.organizeImports": "explicit"
    }
  },
  ```

ps48 commented on src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts:
```diff
@@ -378,6 +380,10 @@ export class IndexPattern implements IIndexPattern {
+  setFieldLoadingStatus = (status: boolean) => {
+    return (this.areFieldsLoading = status);
+  };
+
```
Updated here: [6d6a9c6](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/6d6a9c6b6100e4ca696875e237fe793c07fd84a8)

ps48 commented on src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts:
```diff
@@ -92,25 +93,32 @@ export class DatasetService {
+          title: dataset.title,
+          timeFieldName: dataset.timeFieldName,
+          fields: fetchedFields ?? ({} as IndexPatternFieldMap),
+          areFieldsLoading: true,
```
Not anymore, it used meta flag for aync loading instead of always being true

ps48 commented on src/plugins/data/common/datasets/types.ts:
```diff
@@ -28,6 +28,8 @@ export interface DataSourceMeta {
   name?: string;
   /** Optional session ID for faster responses when utilizing async query sources */
   sessionId?: string;
+  /** Optional requiresTimeFilter determines if a time filter is needed */
```
Updated here: [6d6a9c6](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/6d6a9c6b6100e4ca696875e237fe793c07fd84a8)

joshuali925 commented on src/plugins/data/common/datasets/types.ts:
```diff
@@ -28,6 +28,8 @@ export interface DataSourceMeta {
   /** Optional session ID for faster responses when utilizing async query sources */
   sessionId?: string;
+  /** Optional supportsTimeFilter determines if a time filter is needed */
+  supportsTimeFilter?: boolean;
```
can this be controlled based on whether `timeField` is present in the dataset rather than additional flag?

ps48 replied:
Today even if `timeField` is present we don't have support for S3 datasources to allow time filter injection in the query. This is surely where we want to go towards in future though.

ps48 replied:
Some older conversations around this: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8337#discussion_r1779017573, https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8337#discussion_r1786959825

joshuali925 commented on src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts:
```diff
@@ -91,29 +92,54 @@ export class DatasetService {
+        const temporaryIndexPattern = await this.indexPatterns?.create(spec, true);
+
+        // Load schema asynchronously if it's an async index pattern
+        if (asyncType && temporaryIndexPattern) {
```
does async index patterns mean the metadata like field schema is not stored in the index pattern saved object?

ps48 replied:
Async index patterns means that the fields are lazy loaded; i.e. the temporary index pattern is created with empty fields and when the fields are loaded in the index pattern gets hydrated with the fields.

joshuali925 commented on src/plugins/data/public/ui/dataset_selector/configurator.tsx:
```diff
@@ -80,6 +83,11 @@ export const Configurator = ({
     };
 
+    if (baseDataset?.dataSource?.meta?.supportsTimeFilter === false) {
+      setTimeFields([]);
```
this `[]` creates a different object every time when called, and will trigger a re-render. might be better to add a guard
```suggestion
      if (timeFields.length > 0) setTimeFields([]);
```

ps48 replied:
Updated here: [53198c4](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/53198c40ecdcea449979f3eb51eee7782300cb54)

joshuali925 commented on src/plugins/discover/public/application/components/sidebar/discover_sidebar.tsx:
```diff
@@ -231,11 +231,12 @@ export function DiscoverSidebar(props: DiscoverSidebarProps) {
             paddingSize="none"
           >
-            {fields.length > 0 && (
+            {(fields.length > 0 || selectedIndexPattern.fieldsLoading) && (
```
if loading and `fields` list is empty, what does UI show? should it indicate that it's loading?

ps48 replied:
If loading and fields list is empty -> The UI shows the sidebar in the loading state.

joshuali925 commented on src/plugins/discover/public/application/components/sidebar/discover_sidebar.tsx:
```diff
@@ -312,6 +313,7 @@ const FieldList = ({
         size="xs"
         className="dscSideBar_fieldGroup"
         aria-label={title}
+        isLoading={selectedIndexPattern.fieldsLoading ?? false}
```
```suggestion
        isLoading={selectedIndexPattern.fieldsLoading}
```
or coerce to boolean `!!selectedIndexPattern.fieldsLoading`?

ps48 replied:
Coerce makes sense: [53198c4](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/53198c40ecdcea449979f3eb51eee7782300cb54)

joshuali925 commented on src/plugins/query_enhancements/common/types.ts:
```diff
@@ -36,3 +36,36 @@ export interface QueryStatusOptions<T> {
+export interface SQLQueryResponse {
+  status: string;
+  schema: Array<{ name: string; type: string }>;
+  datarows: Array<Array<string | null>>;
```
should it be
```suggestion
  datarows: Array<Array<unknown>>;
```

i think columns in SQL response is not always string?

ps48 replied:
[53198c4](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/53198c40ecdcea449979f3eb51eee7782300cb54)

joshuali925 commented on src/plugins/query_enhancements/public/datasets/s3_type.ts:
```diff
@@ -271,3 +288,105 @@ const fetchTables = async (http: HttpSetup, path: DataStructure[]): Promise<Data
+    const sessionId = (dataset.dataSource?.meta as DataStructureCustomMeta).sessionId;
+    const response = await http.fetch({
+      method: 'POST',
+      path: trimEnd(`${API.DATA_SOURCE.ASYNC_JOBS}`),
```
nit.
```suggestion
      path: trimEnd(API.DATA_SOURCE.ASYNC_JOBS),
```

ps48 replied:
[53198c4](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/53198c40ecdcea449979f3eb51eee7782300cb54)

joshuali925 commented on src/plugins/query_enhancements/public/datasets/s3_type.ts:
```diff
@@ -271,3 +288,105 @@ const fetchTables = async (http: HttpSetup, path: DataStructure[]): Promise<Data
+      fetchStatus: () =>
+        http.fetch({
+          method: 'GET',
+          path: trimEnd(`${API.DATA_SOURCE.ASYNC_JOBS}`),
```
```suggestion
          path: trimEnd(API.DATA_SOURCE.ASYNC_JOBS),
```

ps48 replied:
Updated here: [53198c4](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8609/commits/53198c40ecdcea449979f3eb51eee7782300cb54)

virajsanghvi commented on src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts:
```diff
@@ -91,29 +92,57 @@ export class DatasetService {
+            .fetchFields(dataset, services)
+            .then((fields) => {
+              temporaryIndexPattern.fields.replaceAll([...fields]);
+              this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
```
Handling of this request should ideally be tested. Can we follow up on this?

virajsanghvi commented on src/plugins/query_enhancements/public/datasets/s3_type.ts:
```diff
@@ -271,3 +288,105 @@ const fetchTables = async (http: HttpSetup, path: DataStructure[]): Promise<Data
+      path: trimEnd(API.DATA_SOURCE.ASYNC_JOBS),
+      body: JSON.stringify({
+        lang: 'sql',
+        query: `DESCRIBE TABLE ${dataset.title}`,
```
does this need to be escaped in anyway or can we trust dataset metadata?

joshuali925 reviewed (approved):
didn't pull down to test but from code perspective lgtm

virajsanghvi reviewed (approved):
Don't have time to pull this down to test, but approving assuming bugs Ashwin brought up are resolved (which appear to be the case from comments)

Other comments are not blocking

ashwin-pc reviewed (changes requested):
~When i pulled it down im seeing issues with the state. Blocking so that we dont accidentally merge it in the meantime~

My bad. Still had the older commits when i rviewed. When i pulled down the latest changes, it works

ashwin-pc commented:
https://github.com/user-attachments/assets/a7af9dd9-e674-44f5-833c-cbbaf6a88fef

Noticed one bug, but wont block it because it can come in a fast follow
